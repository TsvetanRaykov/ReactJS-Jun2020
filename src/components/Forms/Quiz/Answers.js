import React from 'react'
import Single from './AnswerTypes/Single'
import Multiple from './AnswerTypes/Multiple'
import Open from './AnswerTypes/Open'

const Answers = (props) => {
	const { answers, questionType, updateAnswers } = props

	return (
		<>
			{questionType === 'single' && (
				<Single answers={answers} updateAnswers={updateAnswers} />
			)}
			{questionType === 'multiple' && (
				<Multiple answers={answers} updateAnswers={updateAnswers} />
			)}
			{questionType === 'open' && (
				<Open answers={answers} updateAnswers={updateAnswers} />
			)}
		</>
	)
}

export default Answers
