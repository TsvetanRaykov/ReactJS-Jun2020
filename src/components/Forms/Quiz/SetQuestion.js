import React, { useState, useContext, useEffect, Fragment } from 'react'
import {
	Paper,
	Button,
	FormControl,
	Input,
	InputLabel,
	withStyles,
	Box,
	Dialog,
	DialogActions,
	DialogContent,
	createTheme,
	ThemeProvider,
	TextField,
} from '@material-ui/core'
import Answers from './Answers'
import { Add, Cancel, Delete } from '@material-ui/icons'
import { green } from '@material-ui/core/colors'
import Context from '../../../Context'
import ModalDialog from '../../shared/ModalDialog'

const styles = (theme) => ({
	submit: {
		margin: theme.spacing(),
	},
	paper: {
		marginTop: theme.spacing(),
		display: 'flex',
		flexDirection: 'column',
		alignItems: 'center',
		'& form': {
			width: '90%',
		},
	},
	dialog: { width: '90%' },
	button: {
		margin: theme.spacing(1),
	},
})

const theme = createTheme({
	palette: {
		primary: {
			main: green[500],
			contrastText: '#fff',
		},
	},
})

const SetQuestionForm = (props) => {
	const { classes, formClose, newQuestionType, activeQuestionIndex } = props
	const {
		quiz: { questions },
		updateQuiz,
	} = useContext(Context)

	const [open, setOpenAnswerAdd] = useState(false)
	const [question, setQuestion] = useState(
		questions[activeQuestionIndex]?.question || ''
	)

	const [newAnswer, setNewAnswer] = useState('')
	const [answers, setAnswers] = useState([])

	const [modalDialog, setModalDialog] = useState({
		title: '',
		message: '',
		open: false,
		handleYes: () => {},
		handleNo: () => {},
	})

	const questionType = () =>
		questions[activeQuestionIndex]?.type || newQuestionType || 'single'

	const onSaveQuestionClick = () => {
		const newQuestions = questions.slice(0)
		if (activeQuestionIndex >= 0) {
			newQuestions[activeQuestionIndex] = {
				...newQuestions[activeQuestionIndex],
				question,
				answers,
			}
		} else {
			newQuestions.push({ question, answers, type: questionType() })
		}

		updateQuiz({ questions: newQuestions })
		formClose()
	}

	useEffect(() => {
		const ansList = questions[activeQuestionIndex]?.answers || []
		setAnswers(() => ansList.map((a) => Object.assign({}, a)))
		setQuestion(questions[activeQuestionIndex]?.question)
		// eslint-disable-next-line react-hooks/exhaustive-deps
	}, [activeQuestionIndex])

	const hasCorrectAnswer = () => {
		return answers.some((ans) => ans.correct === true)
	}

	const onCancelClick = () => {
		formClose()
	}

	const onDeleteClick = () => {
		setModalDialog({
			title: 'Are you sure?',
			handleYes: () => {
				setModalDialog({ open: false })
				const newQuestions = questions.slice(0)
				newQuestions.splice(activeQuestionIndex, 1)
				updateQuiz({ questions: newQuestions })
				formClose()
			},
			handleNo: () => {
				setModalDialog({ open: false })
			},
			open: true,
		})
	}

	const handleClickOpen = () => {
		setOpenAnswerAdd(true)
	}

	const handleClose = () => {
		if (newAnswer) {
			setAnswers([
				...answers.filter((a) => a.text !== newAnswer),
				{ text: newAnswer, correct: questionType() === 'open' },
			])
			setNewAnswer('')
		}
		setOpenAnswerAdd(false)
	}

	const updateAnswers = (updated) => {
		setAnswers(updated)
	}

	return (
		<Fragment>
			<Paper className={classes.paper}>
				<form onSubmit={(e) => e.preventDefault() && false}>
					<FormControl margin='normal' required fullWidth>
						<InputLabel htmlFor='question'>
							Question ({questionType()} answer
							{questionType() === 'multiple' ? 's' : ''})
						</InputLabel>
						<Input
							id='question'
							name='question'
							autoComplete='off'
							multiline
							autoFocus
							value={question}
							onChange={(e) => setQuestion(e.target.value)}
						/>
					</FormControl>
					<ThemeProvider theme={theme}>
						<Box display='flex' justifyContent='flex-end' flexDirection='row'>
							<Button
								className={classes.button}
								type='button'
								variant='contained'
								color='primary'
								onClick={handleClickOpen}
								startIcon={<Add />}
							>
								Answer
							</Button>
						</Box>
					</ThemeProvider>
					<Box display='flex' justifyContent='space-around' flexDirection='row'>
						<Answers
							answers={answers}
							updateAnswers={updateAnswers}
							questionType={questionType()}
						/>
					</Box>
					<Dialog
						open={open}
						onClose={handleClose}
						aria-labelledby='answer-form-dialog'
						fullWidth={true}
					>
						<DialogContent>
							<TextField
								autoFocus
								margin='dense'
								id='answer'
								label='Answer'
								type='text'
								fullWidth
								multiline
								value={newAnswer}
								onChange={(e) => {
									setNewAnswer(e.currentTarget.value)
								}}
							/>
						</DialogContent>
						<DialogActions>
							<Button onClick={handleClose} color='primary'>
								Cancel
							</Button>
							<Button onClick={handleClose} color='primary'>
								Ok
							</Button>
						</DialogActions>
					</Dialog>
					<Button
						type='submit'
						disabled={!hasCorrectAnswer()}
						variant='contained'
						color='primary'
						onClick={onSaveQuestionClick}
						className={classes.submit}
					>
						Save Question
					</Button>
					<Button
						className={classes.button}
						type='button'
						variant='outlined'
						color='secondary'
						onClick={onCancelClick}
						startIcon={<Cancel />}
					>
						Cancel
					</Button>
					{activeQuestionIndex >= 0 && (
						<Button
							className={classes.button}
							type='button'
							variant='contained'
							color='secondary'
							onClick={onDeleteClick}
							startIcon={<Delete />}
						>
							Delete
						</Button>
					)}
				</form>
			</Paper>
			<ModalDialog {...modalDialog} />
		</Fragment>
	)
}

export default withStyles(styles)(SetQuestionForm)
