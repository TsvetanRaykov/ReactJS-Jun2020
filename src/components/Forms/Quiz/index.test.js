import React from 'react'
import renderer from 'react-test-renderer'
import TestWrapper from '../../../utils/test-wrapper'
import Answers from './Answers'
import SetQuestion from './SetQuestion'

jest.mock('react', () => ({
	...jest.requireActual('react'),
	useLayoutEffect: jest.requireActual('react').useEffect,
}))

describe('<Answers />', () => {
	it('should match the snapshot', () => {
		const wrapper = renderer.create(
			<TestWrapper>
				<Answers selectedValue={jest.fn()} answers={[]} />
			</TestWrapper>
		)
		expect(wrapper).toMatchSnapshot()
	})
})

describe('<SetQuestion />', () => {
	it('should match the snapshot', () => {
		const wrapper = renderer.create(
			<TestWrapper>
				<SetQuestion />
			</TestWrapper>
		)
		expect(wrapper).toMatchSnapshot()
	})
})
