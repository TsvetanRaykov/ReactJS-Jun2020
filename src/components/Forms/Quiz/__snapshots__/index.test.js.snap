// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Answers /> should match the snapshot 1`] = `null`;

exports[`<SetQuestion /> should match the snapshot 1`] = `
Array [
  <div
    className="MuiPaper-root SetQuestionForm-paper-2 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required"
          data-shrink={false}
          htmlFor="question"
        >
          Question (
          single
           answer
          
          )
          <span
            aria-hidden={true}
            className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
            â€‰
            *
          </span>
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-multiline MuiInput-multiline"
          onClick={[Function]}
        >
          <textarea
            aria-invalid={false}
            autoComplete="off"
            autoFocus={true}
            className="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline MuiInput-inputMultiline"
            disabled={false}
            id="question"
            name="question"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={true}
            rows={1}
            style={
              Object {
                "height": undefined,
                "overflow": null,
              }
            }
            value=""
          />
          <textarea
            aria-hidden={true}
            className="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline MuiInput-inputMultiline"
            readOnly={true}
            style={
              Object {
                "height": 0,
                "left": 0,
                "overflow": "hidden",
                "position": "absolute",
                "top": 0,
                "transform": "translateZ(0)",
                "visibility": "hidden",
              }
            }
            tabIndex={-1}
          />
        </div>
      </div>
      <div
        className="MuiBox-root-5 MuiBox-root-6"
      >
        <button
          className="MuiButtonBase-root-34 MuiButton-root-7 MuiButton-contained-15 SetQuestionForm-button-4 MuiButton-containedPrimary-16"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label-8"
          >
            <span
              className="MuiButton-startIcon-29 MuiButton-iconSizeMedium-32"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root-35"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                />
              </svg>
            </span>
            Answer
          </span>
        </button>
      </div>
      <div
        className="MuiBox-root MuiBox-root-44"
      >
        <div
          className="MuiFormControl-root MuiFormControl-fullWidth"
        >
          <div
            aria-label="quiz"
            className="MuiFormGroup-root"
            role="radiogroup"
          />
        </div>
      </div>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained SetQuestionForm-submit-1 MuiButton-containedPrimary Mui-disabled Mui-disabled"
        disabled={true}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="submit"
      >
        <span
          className="MuiButton-label"
        >
          Save Question
        </span>
      </button>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-outlined SetQuestionForm-button-4 MuiButton-outlinedSecondary"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          <span
            className="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
              />
            </svg>
          </span>
          Cancel
        </span>
      </button>
    </form>
  </div>,
  <div />,
]
`;
